batch/batch/driver/instance_collection/pool.py

```diff
@@ -536,65 +536,49 @@ async def schedule_loop_body(self):
+                SCHEDULING_LOOP_JOBS.labels(pool_name=self.pool.name, region=region, scheduled=True).set(0)
+                SCHEDULING_LOOP_JOBS.labels(pool_name=self.pool.name, region=region, scheduled=False).set(0)
             return should_wait
         user_share = {
             user: max(int(300 * resources['allocated_cores_mcpu'] / total + 0.5), 20)
             for user, resources in user_resources.items()
         }
 
-        async def user_runnable_jobs(user, remaining):
-            async for batch in self.db.select_and_fetchall(
-                '''
-SELECT batches.id, batches_cancelled.id IS NOT NULL AS cancelled, userdata, user, format_version
-FROM batches
-LEFT JOIN batches_cancelled
-       ON batches.id = batches_cancelled.id
-WHERE user = %s AND `state` = 'running';
-''',
-                (user,),
-                "user_runnable_jobs__select_running_batches",
-            ):
-                async for record in self.db.select_and_fetchall(
-                    '''
-SELECT job_id, spec, cores_mcpu
-FROM jobs FORCE INDEX(jobs_batch_id_state_always_run_inst_coll_cancelled)
-WHERE batch_id = %s AND state = 'Ready' AND always_run = 1 AND inst_coll = %s
-LIMIT %s;
-''',
-                    (batch['id'], self.pool.name, remaining.value),
-                    "user_runnable_jobs__select_ready_always_run_jobs",
-                ):
-                    record['batch_id'] = batch['id']
-                    record['userdata'] = batch['userdata']
-                    record['user'] = batch['user']
-                    record['format_version'] = batch['format_version']
-                    yield record
-                if not batch['cancelled']:
-                    async for record in self.db.select_and_fetchall(
-                        '''
-SELECT job_id, spec, cores_mcpu
+        async def user_runnable_jobs(user, share):
+            async for record in self.db.select_and_fetchall(
+                f'''
+SELECT user, jobs.batch_id, jobs.job_id, cores_mcpu, spec, userdata, format_version,
+  JSON_ARRAYAGG(region) AS regions,
+  COUNT(region) AS n_regions,
+  ROW_NUMBER() OVER (ORDER BY batch_id ASC, job_id ASC) DIV {share} AS scheduling_iteration
 FROM jobs FORCE INDEX(jobs_batch_id_state_always_run_cancelled)
-WHERE batch_id = %s AND state = 'Ready' AND always_run = 0 AND inst_coll = %s AND cancelled = 0
-LIMIT %s;
+LEFT JOIN batches ON jobs.batch_id = batches.id
+LEFT JOIN batches_cancelled ON batches.id = batches_cancelled.id
+LEFT JOIN job_regions ON jobs.batch_id = job_regions.batch_id AND jobs.job_id = job_regions.job_id
+LEFT JOIN region_ids ON job_regions.region_id = region_ids.region_id
+WHERE user = %s AND batches.`state` = 'running' AND jobs.state = 'Ready' AND (always_run OR batches_cancelled.id IS NULL) AND inst_coll = %s
+GROUP BY jobs.batch_id, jobs.job_id
+ORDER BY scheduling_iteration, user, -n_regions DESC, JSON_UNQUOTE(regions), jobs.batch_id, jobs.job_id
+LIMIT {share * JOB_QUEUE_SCHEDULING_WINDOW_SECONDS};
```

Comment:
Seems like the semantics of the scheduler here have changed quite a bit, and it's hard for me to tell in which ways. Can you elaborate on what the semantic changes in here are? Is there a difference in what gets scheduled between this approach and the one where we first select batches and then select jobs from those batches? Also, why do we care about the scheduling_iteration and the scheduling window here? It's all one scheduling loop, shouldn't we just take up to the user's `share` of jobs?

Reply:
Ah, the second question is a bug and it should just be the `share` and you're right that makes scheduling iteration irrelevant as well as the user in the order by.

The semantics of the scheduler have changed in that we don't necessarily schedule in order of batch id, always_run, and job id and instead try and optimize for packing jobs into instances from the strictest requirements (n_regions = 1) to the loosest requirements (n_regions = 0 or any region) in order to avoid putting a job that has no preference into slots where the jobs do need the specific region.

I think we don't care the order within a user that jobs within a batch and earlier batches get scheduled first. That was more a consequence of how the query was written then a deliberate choice.